clippy_lints/src/misc_early.rs

```diff
@@ -391,91 +390,84 @@ impl EarlyLintPass for MiscEarlyLints {
 
 impl MiscEarlyLints {
     fn check_lit(self, cx: &EarlyContext<'_>, lit: &Lit) {
-        if_chain! {
-            if let LitKind::Int(value, ..) = lit.node;
-            if let Some(src) = snippet_opt(cx, lit.span);
-            if let Some(firstch) = src.chars().next();
-            if char::to_digit(firstch, 10).is_some();
-            then {
-                let mut prev = '\0';
-                for (idx, ch) in src.chars().enumerate() {
-                    if ch == 'i' || ch == 'u' {
-                        if prev != '_' {
-                            span_lint_and_sugg(
-                                cx,
-                                UNSEPARATED_LITERAL_SUFFIX,
-                                lit.span,
-                                "integer type suffix should be separated by an underscore",
-                                "add an underscore",
-                                format!("{}_{}", &src[0..idx], &src[idx..]),
-                                Applicability::MachineApplicable,
-                            );
-                        }
-                        break;
-                    }
-                    prev = ch;
+        if let LitKind::Int(value, lit_int_type) = lit.node {
+            if let Some(src) = snippet_opt(cx, lit.span) {
+                let suffix = match lit_int_type {
+                    LitIntType::Signed(ty) => ty.ty_to_string(),
+                    LitIntType::Unsigned(ty) => ty.ty_to_string(),
+                    LitIntType::Unsuffixed => "",
+                };
+
+                let maybe_last_sep_idx = src.len() - suffix.len() - 1;
+                // Do not suggest when literal is unsuffixed.
+                if !suffix.is_empty() && src.as_bytes()[maybe_last_sep_idx] != b'_' {
+                    span_lint_and_sugg(
+                        cx,
+                        UNSEPARATED_LITERAL_SUFFIX,
+                        lit.span,
+                        "integer type suffix should be separated by an underscore",
+                        "add an underscore",
+                        format!("{}_{}", &src[..=maybe_last_sep_idx], suffix),
+                        Applicability::MachineApplicable,
+                    );
                 }
+
                 if src.starts_with("0x") {
                     let mut seen = (false, false);
-                    for ch in src.chars() {
+                    for ch in src.as_bytes()[2..=maybe_last_sep_idx].iter() {
                         match ch {
-                            'a' ..= 'f' => seen.0 = true,
-                            'A' ..= 'F' => seen.1 = true,
-                            'i' | 'u'   => break,   // start of suffix already
-                            _ => ()
+                            b'a'..=b'f' => seen.0 = true,
+                            b'A'..=b'F' => seen.1 = true,
+                            _ => {},
                         }
                     }
                     if seen.0 && seen.1 {
```

Comment:
This `if` could be moved inside the `for` loop, with a `break` added to it, for slightly better performance.